### PR TITLE
Resolved ALIBI bias regression due to porting flat PA

### PR DESCRIPTION
**Pull request by tannervoas742**
**Requires associated changes on [vllm-hpu-extension PR](https://github.com/HabanaAI/vllm-hpu-extension/pull/34)**

Changes:
- Added back alibi biases to decode stage.
- Optimized ALiBI memory usage.
  - Added environment variable "VLLM_PROMPT_ALIBI_MAX_SEQ_LEN" to allow
    large models to run with restricted prompt lengths.
  - Prompt biases instantiated once in __init__ rather than each
    forward.
  - Prompt and decode biases are shared across encoder/decoder layers.
- Added environment variable "VLLM_ALIBI_USE_FLOAT32_BIASES" to resolve
  accuracy issue on long sequences.
- Updated jais, mpt, falcon, baichuan, and bloom to work with ALiBI.
  - Due to bloom's 176B parameter size I was unable to test this model.
    Its changes are the simplest though.
- Works in lazy and eager mode.
- ALiBI is restricted to "VLLM_PROMPT_USE_FUSEDSDPA=false", and
  "VLLM_CONTIGUOUS_PA=true".
- Add position offsets to improve quality on BS > 1 with sequences of
  varying length.
- BS > 1 may have accuracy issues if on FW < 1.19.0. This is due to
  limitation in softmax. Resolved on FW >= 1.19.0.
- NTT patch for GQA

Co-authored-by: Tanner Voas <tanner.voas@intel.com>
Co-authored-by: Haihao Xiang <haihao.xiang@intel.com>
Signed-off-by: Tanner Voas <tanner.voas@intel.com>

### CODE REVIEW

**Comment by tannervoas742** [comment]
@itaraban @madamczykhabana @kzawora-intel has anyone gotten a chance to review this PR and the associated one on [vllm-hpu-extension](https://github.com/HabanaAI/vllm-hpu-extension/pull/34). I just pushed out a significant update that minimizes changes to non-alibi code sections. It also has significant accuracy and memory optimization changes.

With the current changes ALiBi is now fully functional as long as FW >= 1.19.0 is being used.

Please help review. Any feedback would be appreciated.

**Review by itaraban — APPROVED** [review]

File: vllm/worker/hpu_model_runner.py [hunk 1]
```diff
@@ -1180,6 +1192,30 @@ def _prepare_decode(
                                      slot_mapping=slot_mapping,
                                      lora_ids=lora_ids)
 
+    def _compute_alibi_block(self, block_tables, seq_lens, num_blocks):
+        # Create intermediary and output strctures
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:1196]: Please fix typo as per static code analisys.

File: vllm/attention/backends/hpu_attn.py [hunk 2]
```diff
@@ -125,6 +125,8 @@ def __init__(
         kv_cache_dtype: str,
```
Comment (michalkuligowski) [inline vllm/attention/backends/hpu_attn.py:125]: please modify requirements-hpu.txt to point to apropriate vllm-hpu-extension commit (0766759), fix static code analisys issues and we can merge

**Review by michalkuligowski — CHANGES_REQUESTED** [review]

**Comment by tannervoas742** [comment]
@michalkuligowski I have fixed the static code analysis issue as well as updated requirements-hpu.txt

**Review by michalkuligowski — APPROVED** [review]

**Review by michalkuligowski — CHANGES_REQUESTED** [review]
@tannervoas742  there are still some issues detected, please check (you can try runing format.sh script):
Error: vllm/attention/layer.py:99: error: Too many arguments for "AttentionImpl"  [call-arg]
Error: vllm/attention/backends/hpu_attn.py:279: error: Value of type "Optional[Any]" is not indexable  [index]
Error: vllm/attention/backends/hpu_attn.py:291: error: Item "None" of "Optional[Any]" has no attribute "unsqueeze"  [union-attr]

**Comment by tannervoas742** [comment]
@michalkuligowski I see the issues now. I wasn't sure where to view the static code analysis report, but found it. I pushed out an update. Waiting for the code analysis to run again. Will reply here when it's finished and ready for re-review.

**Comment by tannervoas742** [comment]
@itaraban @michalkuligowski I have updated the PR and ran the script in tools/mypy.sh which passes locally. I also tested the updated version with various ALiBi and non-alibi models. Please re-review. I opened the extension PR again as well. https://github.com/HabanaAI/vllm-hpu-extension/pull/60

File: vllm/worker/hpu_model_runner.py [hunk 3]
```diff
@@ -1130,6 +1137,13 @@ def _prepare_decode(
         block_groups = padding_fn(block_groups, -1)
         block_usage = padding_fn(block_usage, 1)
 
+        alibi_blocks = None
+        if self.use_alibi:
+            alibi_blocks = self._compute_alibi_block(block_tables, seq_lens,
+                                                     len(block_groups))
+            alibi_blocks = alibi_blocks.to(  # type: ignore
```
Comment (kwisniewski98) [inline vllm/worker/hpu_model_runner.py:1144]: Can't we just create tensors on specific devices instead of doing `.to`? I think that this can cause graph breaks in lazy, but I might be wrong

**Review by kwisniewski98 — CHANGES_REQUESTED** [review]
The biggest issue I have right now is that modifying any file that isn't hpu specific (models, attention backends) will cause it to be hard/impossible to upstream. I didn't want to repeat comment for each file, but I think that changes should be removed from all of them.

File: vllm/worker/hpu_model_runner.py (hunk 3 above)
Comment (tannervoas742) [reply vllm/worker/hpu_model_runner.py:1144]: This format is used in later code sections (about 20 lines down). I believe the goal is to compute the structures on the CPU and then asynchronously move to HPU. The exact commit this format was adopted is here. https://github.com/HabanaAI/vllm-fork/commit/7c5038cc0e88b8e2ea027c9f2be4845c2a7f67ad

File: vllm/model_executor/models/baichuan.py [hunk 4]
```diff
@@ -372,6 +390,7 @@ def __init__(
         self.lora_config = lora_config
 
         self.quant_config = quant_config
+        self.use_alibi = position_embedding == "ALIBI"
```
[the remark this answers is not in the document]
Comment (tannervoas742) [reply vllm/model_executor/models/baichuan.py:393]: I was unaware we are not able to make any modifications to even the model definitions. This will be very hard to implement the optimizations (layer sharing) that have been implemented here.

As for this use_alibi is used in hpu_model_runner.py to determine if alibi_blocks are needed or not.

I can remove these changes from the model definitions and move them to hpu_model_runner.py, but doing so there will be more tedious. I understand why we want to avoid modifying non-hpu files though.

**Comment by tannervoas742** [comment]
@kwisniewski98 I refined this PR with only hpu files being changed. I also have rebased this and the extension PR (https://github.com/HabanaAI/vllm-hpu-extension/pull/60) off the latests mains. Tested with several ALiBi and non-alibi models. And local mypy runs showed no new mypy errors.

Please help re-review.

File: vllm/attention/backends/hpu_attn.py [hunk 5]
```diff
@@ -157,12 +166,49 @@ def __init__(
...
                 f"Head size {head_size} is not supported by PagedAttention. "
                 f"Supported head sizes are: {suppored_head_sizes}.")
 
+    def _maybe_init_alibi_biases(
+        self,
+        max_seq_len: int = 4096,
+        prev_attn: Optional[torch.nn.Module] = None,
+    ) -> None:
+        # Set upper bound on sequence length
+        self.max_seq_len_upper = int(
```
Comment (kwisniewski98) [inline vllm/attention/backends/hpu_attn.py:187]: I don't think that this variable has to be object's field. Please remove `self.`

**Review by kwisniewski98 — COMMENTED** [review]
Just last small comment. We will merge https://github.com/HabanaAI/vllm-hpu-extension/pull/70 probably tomorrow, after that you will have to change sha of vllm-hpu-extension in requirements-hpu.txt

**Comment by tannervoas742** [comment]
> Just last small comment. We will merge [HabanaAI/vllm-hpu-extension#70](https://github.com/HabanaAI/vllm-hpu-extension/pull/70) probably tomorrow, after that you will have to change sha of vllm-hpu-extension in requirements-hpu.txt

Understood. I fixed the small issue you mentioned. Will update this PR with the extension sha after that has merged.

**Review by michalkuligowski — CHANGES_REQUESTED** [review]
Please fix the conflicts and static code analisys issues

**Comment by tannervoas742** [comment]
@michalkuligowski @kwisniewski98 conflicts have been resolved. Yapf and ruff issues should also be resolved now.

File: vllm/attention/backends/hpu_attn.py [hunk 6]
```diff
@@ -223,27 +269,46 @@ def forward(
...
-                        attn_bias = attn_bias.tile(
-                            (1, self.num_kv_heads, 1, 1))
-                        attn_bias.add_(position_bias)
+                    position_bias = None
+                    if (self.prompt_position_bias is not None
+                            and self.alibi_slopes is not None):
+                        if self.max_seq_len >= max(attn_bias.size(-2),
+                                                   attn_bias.size(-1)):
+                            # Using pre-computed prompt_position_bias subset.
+                            position_bias = self.prompt_position_bias[:, :,
```
Comment (mswiniarsk) [inline vllm/attention/backends/hpu_attn.py:286]: It is hardly readable, could you please format the line?

Comment (tannervoas742) [reply vllm/attention/backends/hpu_attn.py:286]: This is how yapf formatted it. I had it formatted differently but then yapf CI was failing since it was suggesting this reformat. I also don't like this format but I applied the yapf patch to resolve the CI failures.

**Review by mswiniarsk — APPROVED** [review]

**Comment by PatrykWo** [comment]
@kwisniewski98 please resolve conflicts before merge.

**Comment by tannervoas742** [comment]
> @kwisniewski98 please resolve conflicts before merge, plus some tests failed.

Resolved merge conflicts. Waiting for tests to re-run.

**Review by kwisniewski98 — APPROVED** [review]

**Comment by tannervoas742** [comment]
> @tannervoas742 Could you please switch for now vllm-hpu-extension version to branch that you've added to check if tests will go through?

Updated the reference. The cpu-test passed when I ran it locally.

**Comment by PatrykWo** [comment]
@tannervoas742 please resolve issue in yapf

yapf.....................................................................Failed
- hook id: yapf
- exit code: 1

**Comment by tannervoas742** [comment]
> @tannervoas742 please resolve issue in yapf
> 
> yapf.....................................................................Failed
> 
> * hook id: yapf
> * exit code: 1

Fixed. I setup my pre-commit hooks locally and they passed now including yapf.

**Comment by tannervoas742** [comment]
> @tannervoas742 please rebase

Rebased

**Comment by tannervoas742** [comment]
Freshly rebased. ALiBi fixes still work. Pre-commit hooks all passed.

**Comment by PatrykWo** [comment]
@tannervoas742 please resolve conflict.

**Comment by tannervoas742** [comment]
> @tannervoas742 please resolve conflict.

Resolved conflicts in this and the extension change.

File: requirements-hpu.txt [hunk 7]
```diff
@@ -8,4 +8,4 @@ pandas
 tabulate
 setuptools>=61
 setuptools-scm>=8
-vllm-hpu-extension @ git+https://github.com/HabanaAI/vllm-hpu-extension.git@3fd0250
+vllm-hpu-extension @ git+https://github.com/tannervoas742/vllm-hpu-extension.git@restore_alibi_for_flat_pa_final
```
[the remark this answers is not in the document]
Comment (tannervoas742) [reply requirements-hpu.txt:11]: Im aware. I will update this PR as soon as the extension merges.

**Comment by kwisniewski98** [comment]
@tannervoas742 hpu extension PR was merged, please change `requirements-hpu.txt`

**Comment by tannervoas742** [comment]
> @tannervoas742 hpu extension PR was merged, please change `requirements-hpu.txt`

Conflicts resolved and extension reference was updated.

**Review by michalkuligowski — APPROVED** [review]
